### ISSUE TITLE
bug: vm.max_map_count exceeded

### ISSUE THREAD

**Issue by antonya86** [solana-labs/solana#27753]
#### Problem

solana-validator reached vm.max_map_count (1mil) and crashed. Looks like a bug. Please, check logs below:

```
[2022-09-08T05:48:09.922161045Z ERROR solana_runtime::append_vec] Failed to map the data file (size: 786432): Cannot allocate memory (os error 12).

                        Please increase sysctl vm.max_map_count or equivalent for your platform.
```

vm.max_map_count was set to 1000000 according to documentation. solana-validator confirmed it within startup:
` vm.max_map_count: recommended=1000000 current=1000000`

Normally solana_validator process consumes around `~450k` mmap but before crash it hit 1mil. We have not faced such issue before and caught this bug only on solana-validator version **v1.10.38**

#### Proposed Solution
fix the bug

**Comment by sakridge** [comment]
cc @jeffwashington

**Comment by jeffwashington** [comment]
Hello @antonya86.
Can you please look for a log line that looks like this:
```
 INFO  solana_runtime::accounts_db] generating index: 8088/420232 slots...
 ```
 I need to know the second number. It is likely the validator didn't make it that far.
 
 If not, please look for the last group of 4 log lines that look like this.
 ```
[2022-09-13T16:08:15.708688468Z INFO  solana_runtime::hardened_unpack] unpacked 33063 entries so far...
[2022-09-13T16:08:15.709260207Z INFO  solana_runtime::hardened_unpack] unpacked 33063 entries so far...
[2022-09-13T16:08:15.709417036Z INFO  solana_runtime::hardened_unpack] unpacked 33063 entries so far...
[2022-09-13T16:08:15.919128465Z INFO  solana_runtime::hardened_unpack] unpacked 33062 entries so far...
 ```
 
I think you have a snapshot that has far too many slots stored in it. There is a known bug in 1.10 and prior that we have a fix for that causes snapshots to contain too many slots.
 
In this case, the most likely remedy is to get a new snapshot.

**Comment by antonya86** [comment]
Hello @jeffwashington 

Just to be clear, the validator crashed **not** on startup, it was up, running and validating ~9 days before it got crashed. 

Another important thing to mention, snapshots were disabled on the solana-validator by adding option `--incremental-snapshot-interval-slots 0`. So, `solana-validator ` didn't make snapshots.

Validator crashed at 5.48 UTC, next we fetched snapshot from the mainnet-beta cluster and started the validator. There were messages you asked about after validator restart:
```
[2022-09-08T05:57:05.743507794Z INFO  solana_runtime::accounts_db] generating index: 24973/605500 slots...
[2022-09-08T05:57:07.743565935Z INFO  solana_runtime::accounts_db] generating index: 109381/605500 slots...
[2022-09-08T05:57:09.743591850Z INFO  solana_runtime::accounts_db] generating index: 171517/605500 slots...
```
```[2022-09-08T05:55:43.353847639Z INFO  solana_runtime::hardened_unpack] unpacked 20651 entries so far...
[2022-09-08T05:55:43.353888300Z INFO  solana_runtime::hardened_unpack] unpacked 20651 entries so far...
[2022-09-08T05:55:43.353913300Z INFO  solana_runtime::hardened_unpack] unpacked 20651 entries so far...
[2022-09-08T05:55:43.354007766Z INFO  solana_runtime::hardened_unpack] unpacked 20651 entries so far...
[2022-09-08T05:55:53.367146434Z INFO  solana_runtime::hardened_unpack] unpacked 42511 entries so far...
[2022-09-08T05:55:53.367199153Z INFO  solana_runtime::hardened_unpack] unpacked 42511 entries so far...
[2022-09-08T05:55:53.367208152Z INFO  solana_runtime::hardened_unpack] unpacked 42511 entries so far...
[2022-09-08T05:55:53.367206862Z INFO  solana_runtime::hardened_unpack] unpacked 42511 entries so far...
[2022-09-08T05:56:03.378424778Z INFO  solana_runtime::hardened_unpack] unpacked 64353 entries so far...
[2022-09-08T05:56:03.378432014Z INFO  solana_runtime::hardened_unpack] unpacked 64353 entries so far...
[2022-09-08T05:56:03.378547820Z INFO  solana_runtime::hardened_unpack] unpacked 64353 entries so far...
[2022-09-08T05:56:03.378621399Z INFO  solana_runtime::hardened_unpack] unpacked 64354 entries so far...
[2022-09-08T05:56:13.382609402Z INFO  solana_runtime::hardened_unpack] unpacked 86638 entries so far...
[2022-09-08T05:56:13.382622678Z INFO  solana_runtime::hardened_unpack] unpacked 86638 entries so far...
[2022-09-08T05:56:13.382636004Z INFO  solana_runtime::hardened_unpack] unpacked 86638 entries so far...
[2022-09-08T05:56:13.382652391Z INFO  solana_runtime::hardened_unpack] unpacked 86637 entries so far...
[2022-09-08T05:56:23.382627822Z INFO  solana_runtime::hardened_unpack] unpacked 108884 entries so far...
[2022-09-08T05:56:23.382682572Z INFO  solana_runtime::hardened_unpack] unpacked 108885 entries so far...
[2022-09-08T05:56:23.382722525Z INFO  solana_runtime::hardened_unpack] unpacked 108890 entries so far...
[2022-09-08T05:56:23.418573581Z INFO  solana_runtime::hardened_unpack] unpacked 108895 entries so far...
[2022-09-08T05:56:33.382629650Z INFO  solana_runtime::hardened_unpack] unpacked 130874 entries so far...
[2022-09-08T05:56:33.382773521Z INFO  solana_runtime::hardened_unpack] unpacked 130864 entries so far...
[2022-09-08T05:56:33.382925074Z INFO  solana_runtime::hardened_unpack] unpacked 130875 entries so far...
[2022-09-08T05:56:33.428523176Z INFO  solana_runtime::hardened_unpack] unpacked 130926 entries so far...
[2022-09-08T05:56:40.507419205Z INFO  solana_runtime::hardened_unpack] unpacked 146676 entries total
[2022-09-08T05:56:40.509868548Z INFO  solana_runtime::hardened_unpack] unpacked 146676 entries total
[2022-09-08T05:56:40.509897679Z INFO  solana_runtime::hardened_unpack] unpacked 146677 entries total
[2022-09-08T05:56:40.517341217Z INFO  solana_runtime::hardened_unpack] unpacked 146677 entries total
[2022-09-08T05:56:44.669681757Z INFO  solana_runtime::hardened_unpack] unpacked 4701 entries total
[2022-09-08T05:56:44.670229869Z INFO  solana_runtime::hardened_unpack] unpacked 4702 entries total
[2022-09-08T05:56:44.670229260Z INFO  solana_runtime::hardened_unpack] unpacked 4701 entries total
[2022-09-08T05:56:44.679475011Z INFO  solana_runtime::hardened_unpack] unpacked 4702 entries total```

**Comment by jeffwashington** [comment]
Ok, then prior to the crash, please look for:
```
 INFO  solana_metrics::metrics] datapoint: calculate_accounts_hash_without_index mark_time_us=2717i ... num_snapshot_storage=394363i num_slots=394363i ...
 ```
 I need to know `num_snapshot-storage` and `num_slots`.

**Comment by jeffwashington** [comment]
I believe this is your issue:
#26375

We have a fix in 1.11 (#27461). We are waiting to make sure it is safe before considering 1.10. In the meantime, it is possible to get a snapshot (or generate state locally) that causes us to accumulate old slots and append vec files, causing us to run out of mapped files.

**Comment by antonya86** [comment]
Here you go:

`[2022-09-08T05:47:43.854872411Z INFO  solana_metrics::metrics] datapoint: calculate_accounts_hash_without_index accounts_scan=2760695i eliminate_zeros=8333576i hash=4401582i hash_time_pre_us=4400352i sort=132505i hash_total=246797461i storage_sort_us=350950i unreduced_entries=269246223i collect_snapshots_us=349574i num_snapshot_storage=994466i num_slots=994466i min_bin_size=3778i max_bin_size=8255i total=16196377i
`

Looks like it is #27461

**Comment by jeffwashington** [comment]
Yes, your validator was accumulating old slots (994,466 vs the desired < 432k), with each slot having at least one open file. This resulted in the os failing to open any more mmaps. Thus, the validator crashes and restarts.
I believe this is fixed by #27461.
